src/core/ddsi/src/ddsi_raweth.c

```diff
@@ -391,6 +407,323 @@ static int joinleave_asm_mcgroup (ddsrt_socket_t socket, int join, const ddsi_lo
   return (rc == DDS_RETCODE_OK) ? 0 : rc;
 }
 
+#elif DDSI_USE_BSD
+
+#define DEFAULT_BUFFER_SIZE 1000000
+
+struct ddsi_vlan_tag {
+  unsigned short tag;
+  unsigned short proto; 
+};
+
+/* The ddsi_raweth_conn_read reads from the bpf file descriptor.
+ * The read copies the contents of the kernel bpf buffer to a buffer maintained
+ * by this raweth transport. The buffer that is returned may contain several ethernet frames.
+ * Each ethernet frame is preceded by a header (bpf_hdr) which contains the following fields:
+ * - struct bpf_ts bh tstamp : timestamp
+ * - uint32_t bh_captlen     : captured length
+ * - uint32_t bh_datalen     : length of the captured frame
+ * - ushort bh_hdrlen        : length of this header including alignment
+ * Each ddsi_raweth_conn_read will return one packet from the buffer. When the buffer has become empty
+ * then the buffer is filled again by reading from the bpf file descriptor.
+ * This bpf_header is provided by the kernel therefore we can trust the contents of these fields and
+ * the manipulations using the field to obtain the next packet in the buffer can be safely done.
+ */
+
+static ssize_t ddsi_raweth_conn_read (struct ddsi_tran_conn * conn, unsigned char * buf, size_t len, bool allow_spurious, ddsi_locator_t *srcloc)
+{
+  ssize_t ret  = 0;
+  dds_return_t rc = DDS_RETCODE_OK;
+  ddsi_raweth_conn_t uc = (ddsi_raweth_conn_t) conn;
+  struct bpf_hdr *bpf_hdr;
+  struct ddsi_ethernet_header *eth_hdr;
+  struct ddsi_vlan_tag *vtag = NULL;
+  char *ptr;
+  (void) allow_spurious;
+
+  ddsrt_mutex_lock (&uc->lock);
+
+  if (uc->avail == 0)
+  {
+    if ((ret = read(uc->m_sock, uc->buffer, uc->buflen)) <= 0)
+    {
+      DDS_CERROR (&conn->m_base.gv->logconfig, "ddsi_raweth_create_conn read failed ... retcode = %"PRIdSIZE"\n", ret);
+      rc = DDS_RETCODE_ERROR;
+      goto error;
+    }
+    uc->avail = ret;
+    uc->bptr = uc->buffer;
+  }
+
+  if (uc->bptr < uc->buffer + uc->avail)
+  {
+    ptr = uc->bptr;
+    bpf_hdr = (struct bpf_hdr *) ptr;
+    ptr += bpf_hdr->bh_hdrlen;
+  
+    eth_hdr = (struct ddsi_ethernet_header *)ptr;
+    ptr += sizeof(*eth_hdr);
+
+    if (bpf_hdr->bh_datalen == bpf_hdr->bh_caplen) 
+    {
+      ret = (ssize_t)(bpf_hdr->bh_datalen - sizeof(struct ddsi_ethernet_header));
+      if (ntohs(eth_hdr->proto) == ETHERTYPE_VLAN)
+      {
+        vtag = (struct ddsi_vlan_tag *)ptr;
+        ptr += sizeof(*vtag);
+        ret -= (ssize_t)sizeof(*vtag);
+      }
+      if ((size_t)ret <= len)
+      {
+        memcpy(buf, ptr, (size_t)ret);
+        if (srcloc)
+          set_locator(srcloc, eth_hdr->smac, ntohs (eth_hdr->proto), (vtag ? vtag->tag : 0));
```

Comment:
Like in the Linux version, on consideration I am somewhat surprised that the tag would be in native byte order already. Perhaps you can double-check?

Reply:
In this case the vtag is of course in network order. I will correct this.